worker/manager/src/mirrors_qa_manager/docker.py

```diff
@@ -0,0 +1,141 @@
+# pyright: strict, reportMissingTypeStubs=false, reportUnknownMemberType=false, reportOptionalSubscript=false, reportUnknownVariableType=false, reportUnknownArgumentType=false
+import time
+from collections.abc import Callable
+from functools import wraps
+from pathlib import Path
+from typing import Any
+
+from docker import DockerClient
+from docker.errors import APIError, ImageNotFound, NotFound
+from docker.models.containers import Container, ExecResult
+from docker.models.images import Image
+
+from mirrors_qa_manager import logger
+from mirrors_qa_manager.settings import Settings
+
+
+def retry(
+    retries: int = Settings.DOCKER_API_RETRIES,
+    interval: int = Settings.DOCKER_API_RETRY_SECONDS,
+):
+    """Retry API calls to the Docker daemon on failure."""
+
+    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
+        @wraps(func)
+        def wrapped(*args: Any, **kwargs: Any):
+            attempt = 0
+            while True:
+                attempt += 1
+                try:
+                    return func(*args, **kwargs)
+                # most 404 errors would be from incorrect image/tag name but
+                # we still want to avoid crashing on 404 due to temporary
+                # docker-hub issues
+                except (ImageNotFound, APIError) as exc:
+                    if attempt <= retries:
+                        logger.error(
+                            f"docker api error for {func.__name__} "
+                            f"(attempt {attempt}): {exc}"
+                        )
+                        time.sleep(interval * attempt)
+                        continue
+                    raise exc
+
+        return wrapped
+
+    return decorator
+
+
+def get_running_container_name() -> str:
+    """Determine the name of this container."""
+    with open("/etc/hostname") as fh:
+        return Path(fh.read().strip()).name
+
+
+def get_container(client: DockerClient, container_name: str) -> Container:
+    return client.containers.get(
+        container_name
+    )  # pyright: ignore [reportGeneralTypeIssues,reportReturnType]
+
+
+@retry()
+def get_docker_client() -> DockerClient:
+    logger.info(f"Connecting to Docker API on {Settings.DOCKER_SOCKET}")
+    client = DockerClient(
+        base_url=f"unix://{Settings.DOCKER_SOCKET}",
+        timeout=Settings.DOCKER_CLIENT_TIMEOUT,
+    )
+    client.ping()
+    logger.info("Connected to Docker API successfully")
+    return client
+
+
+@retry()
+def remove_container(
+    client: DockerClient,
+    container_id: str,
+    *,
+    remove_volumes: bool = False,
+    force: bool = False,
+    not_found_ok: bool = False,
+) -> None:
+    try:
+        container = get_container(client, container_id)
+    except NotFound as exc:
+        if not_found_ok:
+            pass
```

Comment:
You should return here no?

Reply:
Wow, thanks for pointing that out.😅


Couldn't figure out a way to write tests for the docker code